### PR TITLE
make generic test generator to scaffold basic test structure

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
Should be good to go now.

File: bin/jackal-test [hunk 1]
```diff
@@ -2,4 +2,27 @@
...
+require 'jackal'
+
+# This conditional is necessary since the cli parsing has a no-arg
+#   default of printing the help menu (not able to override the default)
+if ARGV.empty?
+  require 'jackal/utils/spec/loader'
+else
+  Bogo::Cli::Setup.define do
+    global_opts = lambda do
+      on :s, 'service_name=', 'Service to test (Eg: Jackal::Mail::Smtp - mail)'
```
Comment (chrisroberts) [inline bin/jackal-test]: change the underscores to dashes in the options `'service_name='` to `'service-name='` and same for module name below

File: lib/jackal/utils/spec/generator.rb [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  module Utils
+    module Spec
+      class Generator < Bogo::Cli::Command
+        TEST_DIRS = 'test/specs/config'
+
+        def initialize(opts, args)
+          super
+
+          @orig_service_name = opts[:service_name]
+          @service_name = @orig_service_name.gsub(/-/, '_')
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]:
``` ruby
@service_name = @orig_service_name.tr('-', '_')
```

File: bin/jackal-test (hunk 1 above)
Comment (ghost) [inline bin/jackal-test]: only issue with that is then you end up getting option params from (cli / slop) that look like `:"service-name"` which is a bit unexpected :neutral_face:

File: lib/jackal/utils/spec/generator.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          @service_class_name = @service_name.split('_').map(&:capitalize).join
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]:
``` ruby
@service_class_name = Bogo::Utility.snake(@service_name)
```

File: lib/jackal/utils/spec/generator.rb [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+require 'fileutils'
+require 'jackal'
+
+module Jackal
+  module Utils
+    module Spec
+      class Generator < Bogo::Cli::Command
+        TEST_DIRS = 'test/specs/config'
+
+        def initialize(opts, args)
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]: change the `opts` name to something else, like `opts_orig`. no need to change later usage of `opts`.

File: lib/jackal/utils/spec/generator.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+          super
+
+          @orig_service_name = opts[:service_name]
+          @service_name = @orig_service_name.gsub(/-/, '_')
+          @service_class_name = @service_name.split('_').map(&:capitalize).join
+
+          @module_name = opts[:module_name]
+          @module_class_name = @module_name.capitalize
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]:
``` ruby
@module_class_name = Bogo::Utility.camel(@module_name)
```

File: lib/jackal/utils/spec/generator.rb [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  end
+
+end
+TEXT
+        end
+
+        private
+
+        def callback_class
+          @callback_class ||= @callback_type.capitalize
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]:
``` ruby
@callback_class ||= Bogo::Utility.camel(@callback_type)
```

File: bin/jackal-test (hunk 1 above)
Comment (chrisroberts) [inline bin/jackal-test]: this fixes that: https://github.com/carnivore-rb/jackal/pull/7/files#r34399386

File: lib/jackal/utils/spec/generator.rb [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+module Jackal
+  module Utils
+    module Spec
+      class Generator < Bogo::Cli::Command
+        TEST_DIRS = 'test/specs/config'
+
+        def initialize(orig_opts, _)
+          super
+
+          @orig_service_name = orig_opts[:"service-name"]
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]:
``` ruby
opts[:service_name]
```

File: lib/jackal/utils/spec/generator.rb [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+          @service_name = Bogo::Utility.snake(@orig_service_name)
+          @service_class_name = @service_name.split('_').map(&:capitalize).join
+
+          @module_name = orig_opts[:"module-name"]
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]:
``` ruby
opts[:module_name]
```

File: lib/jackal/utils/spec/generator.rb [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    module Spec
+      class Generator < Bogo::Cli::Command
+        TEST_DIRS = 'test/specs/config'
+
+        def initialize(orig_opts, _)
+          super
+
+          @orig_service_name = orig_opts[:"service-name"]
+          @service_name = Bogo::Utility.snake(@orig_service_name)
+          @service_class_name = @service_name.split('_').map(&:capitalize).join
```
Comment (chrisroberts) [inline lib/jackal/utils/spec/generator.rb]:
``` ruby
Bogo::Utility.camel(@service_name)
```

File: lib/jackal/utils/spec/generator.rb (hunk 8 above)
Comment (ghost) [inline lib/jackal/utils/spec/generator.rb]: weird, opts is {} here, but `options` works

**Comment by ghost** [comment]
@chrisroberts this good to merge?

**Comment by chrisroberts** [comment]
SHAAAAAAAAA

**Comment by ghost** [comment]
lol, thanks for the review / merge!

**Pull request merged**